File: frontend/src/lib/profileDownloadName.ts
import type { VpnConfig } from '@/types'

export type ProfileFile = VpnConfig['profile_files'][number]

const CLIENT_NAME_RE = /^[a-zA-Z0-9_-]{1,32}$/
const AZ_PROFILE_DIR = /\/(?:openvpn|wireguard|amneziawg)\/antizapret(?:[-/]|$)/

function sanitizeClientName(clientName: string): string {
  const name = clientName.trim()
  return CLIENT_NAME_RE.test(name) ? name : 'client'
}

function parseProfileLocation(path: string): { protocol: string; variant: string } {
  const normalized = path.replace(/\\/g, '/')
  const marker = '/client/'
  const idx = normalized.indexOf(marker)
  if (idx === -1) return { protocol: '', variant: '' }
  const tail = normalized.slice(idx + marker.length).split('/')
  return { protocol: tail[0] ?? '', variant: tail[1] ?? '' }
}

function isAzProfile(variant: string, path: string): boolean {
  if (variant.includes('antizapret')) return true
  return AZ_PROFILE_DIR.test(path.replace(/\\/g, '/'))
}

function openvpnSuffix(variant: string, path: string): string {
  const normalized = `${variant} ${path}`.toLowerCase()
  if (normalized.includes('-udp')) return '-udp'
  if (normalized.includes('-tcp')) return '-tcp'
  return ''
}

function buildProfileDownloadFilename(
  clientName: string,
  file: Pick<ProfileFile, 'protocol' | 'variant' | 'path'>,
): string {
  const safeName = sanitizeClientName(clientName)
  const location = parseProfileLocation(file.path)
  const protocol = (file.protocol || location.protocol).toLowerCase()
  const variant = file.variant || location.variant
  const profilePrefix = isAzProfile(variant, file.path) ? 'AZ' : 'VPN'

  if (protocol === 'openvpn') {
    return `${profilePrefix}-${safeName}${openvpnSuffix(variant, file.path)}.ovpn`
  }
  if (protocol === 'wireguard') {
    return `WG-${profilePrefix}-${safeName}.conf`
  }
  if (protocol === 'amneziawg') {
    return `AWG-${profilePrefix}-${safeName}.conf`
  }
  if (protocol === 'amneziawg2') {
    const pathName = (file.path.split('/').pop() || '').toLowerCase()
    if (pathName.endsWith('-am.conf') || pathName.endsWith('.conf')) {
      return `AWG2-${profilePrefix}-${safeName}.conf`
    }
    if (pathName.endsWith('.vpn')) {
      return `AWG2-${profilePrefix}-${safeName}.vpn`
    }
    if (pathName.includes('vpnuri') || pathName.endsWith('.txt')) {
      return `AWG2-${profilePrefix}-${safeName}-vpnuri.txt`
    }
    return file.path.split('/').pop() || `AWG2-${profilePrefix}-${safeName}.conf`
  }

  const fallback = file.path.split('/').pop()
  return fallback || `${profilePrefix}-${safeName}.txt`
}

export function getProfileDownloadFilename(clientName: string, file: ProfileFile): string {
  return file.download_filename || buildProfileDownloadFilename(clientName, file)
}

export function parseContentDispositionFilename(header: string | null): string | null {
  if (!header) return null
  const utf8Match = /filename\*=UTF-8''([^;]+)/i.exec(header)
  if (utf8Match?.[1]) {
    try {
      return decodeURIComponent(utf8Match[1].trim())
    } catch {
      return utf8Match[1].trim()
    }
  }
  const quotedMatch = /filename="([^"]+)"/i.exec(header)
  if (quotedMatch?.[1]) return quotedMatch[1]
  const plainMatch = /filename=([^;]+)/i.exec(header)
  return plainMatch?.[1]?.trim() ?? null
}

File: backend/app/services/profile_download_name.py
"""Human-friendly download filenames for VPN profile files."""

from __future__ import annotations

import re
from pathlib import PurePosixPath

CLIENT_NAME_RE = re.compile(r"^[a-zA-Z0-9_-]{1,32}$")
AZ_PROFILE_DIR = re.compile(r"/(?:(?:openvpn|wireguard|amneziawg)/antizapret|clients/antizapret)(?:[-/]|$)")


def sanitize_client_name(client_name: str) -> str:
    name = (client_name or "").strip()
    if CLIENT_NAME_RE.match(name):
        return name
    return "client"


def _parse_profile_location(path: str) -> tuple[str, str]:
    parts = PurePosixPath(path).parts
    if "client" in parts:
        idx = parts.index("client")
        if idx + 2 >= len(parts):
            return "", ""
        return parts[idx + 1], parts[idx + 2]
    if "clients" in parts:
        idx = parts.index("clients")
        if idx + 2 >= len(parts):
            return "", ""
        return "amneziawg2", parts[idx + 1]
    return "", ""


def _is_az_profile(*, variant: str, path: str) -> bool:
    if "antizapret" in variant:
        return True
    return bool(AZ_PROFILE_DIR.search(path.replace("\\", "/")))


def _openvpn_suffix(variant: str, path: str) -> str:
    normalized = f"{variant} {PurePosixPath(path).name}".lower()
    if "-udp" in normalized:
        return "-udp"
    if "-tcp" in normalized:
        return "-tcp"
    return ""


def build_profile_download_filename(
    client_name: str,
    *,
    protocol: str = "",
    variant: str = "",
    path: str = "",
) -> str:
    safe_name = sanitize_client_name(client_name)
    proto = (protocol or _parse_profile_location(path)[0]).lower()
    variant = variant or _parse_profile_location(path)[1]
    profile_prefix = "AZ" if _is_az_profile(variant=variant, path=path) else "VPN"

    if proto == "openvpn":
        suffix = _openvpn_suffix(variant, path)
        return f"{profile_prefix}-{safe_name}{suffix}.ovpn"
    if proto == "wireguard":
        return f"WG-{profile_prefix}-{safe_name}.conf"
    if proto == "amneziawg":
        return f"AWG-{profile_prefix}-{safe_name}.conf"
    if proto == "amneziawg2":
        # Primary tunnel profiles are *-am.conf; sidecars (.vpn / vpnuri) keep a distinct name.
        path_name = PurePosixPath(path).name if path else ""
        lower_name = path_name.lower()
        if lower_name.endswith("-am.conf") or lower_name.endswith(".conf"):
            return f"AWG2-{profile_prefix}-{safe_name}.conf"
        if lower_name.endswith(".vpn"):
            return f"AWG2-{profile_prefix}-{safe_name}.vpn"
        if "vpnuri" in lower_name or lower_name.endswith(".txt"):
            return f"AWG2-{profile_prefix}-{safe_name}-vpnuri.txt"
        return path_name or f"AWG2-{profile_prefix}-{safe_name}.conf"

    raw = PurePosixPath(path).name if path else f"{profile_prefix}-{safe_name}.txt"
    return raw or f"{profile_prefix}-{safe_name}.txt"


def enrich_profile_files(client_name: str, files: list[dict[str, str]]) -> list[dict[str, str]]:
    enriched: list[dict[str, str]] = []
    for item in files:
        row = dict(item)
        row["download_filename"] = build_profile_download_filename(
            client_name,
            protocol=item.get("protocol", ""),
            variant=item.get("variant", ""),
            path=item.get("path", ""),
        )
        enriched.append(row)
    return enriched
